openedx/core/djangoapps/notifications/events.py

```diff
@@ -0,0 +1,109 @@
+""" Events for notification app. """
+
+from eventtracking import tracker
+from common.djangoapps.track import contexts
+
+
+NOTIFICATION_PREFERENCES_VIEWED = 'edx.notifications.preferences.viewed'
+NOTIFICATION_GENERATED = 'edx.notifications.generated'
+NOTIFICATION_READ = 'edx.notifications.read'
+NOTIFICATION_PREFERENCES_UPDATED = 'edx.notifications.preferences.updated'
+
+
+def get_user_forums_roles(user, course_id):
+    """
+    Get the user's roles in the course forums.
+    """
+    if course_id:
+        return list(user.roles.filter(course_id=course_id).values_list('name', flat=True))
+    return []
+
+
+def get_user_course_roles(user, course_id):
+    """
+    Get the user's roles in the course.
+    """
+    if course_id:
+        return list(user.courseaccessrole_set.filter(course_id=course_id).values_list('role', flat=True))
+    return []
+
+
+def notification_event_context(user, course_id, notification):
+    return {
+        'user_id': str(user.id),
+        'course_id': str(course_id),
+        'notification_type': notification.notification_type,
+        'notification_app': notification.app_name,
+        'notification_metadata': {
+            'notification_id': notification.id,
+            'notification_content': notification.content,
+        },
+        'user_forum_roles': get_user_forums_roles(user, course_id),
+        'user_course_roles': get_user_course_roles(user, course_id),
+    }
+
+
+def notification_preferences_viewed_event(request, course_id):
+    """
+    Emit an event when a user views their notification preferences.
+    """
+    event_name = NOTIFICATION_PREFERENCES_VIEWED
+    context = contexts.course_context_from_course_id(course_id)
+    with tracker.get_tracker().context(event_name, context):
+        tracker.emit(
+            event_name,
+            {
+                'user_id': str(request.user.id),
+                'course_id': str(course_id),
+                'user_forum_roles': get_user_forums_roles(request.user, course_id),
+                'user_course_roles': get_user_course_roles(request.user, course_id),
+            }
+        )
+
+
+def notification_generated_event(user, notification):
+    """
+    Emit an event when a notification is generated.
+    """
+    event_name = NOTIFICATION_GENERATED
```

Comment:
NOTIFICATION_GENERATED can be directly used not need to create a new variable event_name

Reply:
Have defined the events as constants at the top of the file for better readability

Reply:
why `event_name` variable declaration  is needed here, NOTIFICATION_GENERATED can be directly used .
```python
    def notification_generated_event(user, notification):
    """
    Emit an event when a notification is generated.
    """
    context = contexts.course_context_from_course_id(notification.course_id)
    with tracker.get_tracker().context(NOTIFICATION_GENERATED, context):
        tracker.emit(
            NOTIFICATION_GENERATED,
            notification_event_context(user, notification.course_id, notification)
        )
```

Reply:
Oops, yes. Updating this. Thanks for pointing it out 👍